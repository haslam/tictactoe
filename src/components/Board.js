import React from 'react';
import Cards from './Cards'
import styles from '../styles/board.module.css';
import { resetButton } from '../styles/button.module.css';

export default class Board extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      cards: Array(9).fill(null),
      nextPlay: true,
      win: ''
    }
    this.handlePick = this.handlePick.bind(this);
    this.resetPicks = this.resetPicks.bind(this);
  }
  //Reset the state to init default
  resetPicks() {
    this.setState({
      cards: Array(9).fill(null),
      nextPlay: true,
      win: '',
    })
  }
  /**
   * pick handler for card play
   * @param {int} index Takes in zero index of the array item
   */
  handlePick(index) {
    console.log(index);
    //head back if card already clicked
    if(this.state.cards[index] !== null || this.gameWinner(this.state.cards)) 
      return;

    const reCard = this.state.cards.slice(0); //create a new copy of the cards array so as not to mutate.
    reCard[index] = this.state.nextPlay ? 'X' : 'O'; //set index value to X or O depending on nextPlay boolean value
    this.setState({ 
      cards: reCard,
      nextPlay: !this.state.nextPlay, //flip and set nextPlay
     })
    
  }
  
  /**
   * Calculate winning pick based on matching set rows
   * Takes in the complete array
   * @param {Array} picks returns a single key value of the winning row.
   */
  gameWinner(picks) {
      const rowMix = [
        [0, 1, 2],
        [3, 4, 5],
        [6, 7, 8],
        [0, 3, 6],
        [1, 4, 7],
        [2, 5, 8],
        [0, 4, 8],
        [2, 4, 6],
      ];

      for (let row of rowMix) {
        const [a, b, c] = row;  //destructure array value as a b c

        //check if first value exist and first value is same as second and third [a b c]
        //if so, it's a match
        if (picks[a] && picks[a] === picks[b] && picks[a] === picks[c]) {

          //set the winning pick [ X or O ]
          this.setState(() => ({ win: picks[a] }));

          //reset cards after 3 seconds to allow user read winning pick before clear
          setTimeout(() => {
            this.resetPicks()
          }, 3000);

          return picks[a];
        }
      }
      return false;
    }

    render() {
      return (
        <div>
          <h2 className={styles.winnerText}>
            {this.state.win && `Winner is ${this.state.win}`}
          </h2>
          <Cards
            cards={this.state.cards}
            handlePick={this.handlePick}
          />
          <button className={resetButton} onClick={() => this.resetPicks()}>Reset</button>
        </div>
      )
    }

  }

