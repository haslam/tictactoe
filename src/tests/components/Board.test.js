import React from 'react';
import { shallow } from 'enzyme';
import Board from '../../components/Board';

describe('Board', () => {
  const wrapper = shallow(<Board />);

  it('should render as expected', () => {
    expect(wrapper).toMatchSnapshot();
  })
})