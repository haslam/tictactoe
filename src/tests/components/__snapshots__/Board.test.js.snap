// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board should render as expected 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Board />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h2
          className="winnerText"
        >
          
        </h2>,
        <Cards
          cards={
            Array [
              null,
              null,
              null,
              null,
              null,
              null,
              null,
              null,
              null,
            ]
          }
          handlePick={[Function]}
        />,
        <button
          className="resetButton"
          onClick={[Function]}
        >
          Reset
        </button>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "",
          "className": "winnerText",
        },
        "ref": null,
        "rendered": "",
        "type": "h2",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "cards": Array [
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null,
          ],
          "handlePick": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Reset",
          "className": "resetButton",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "Reset",
        "type": "button",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h2
            className="winnerText"
          >
            
          </h2>,
          <Cards
            cards={
              Array [
                null,
                null,
                null,
                null,
                null,
                null,
                null,
                null,
                null,
              ]
            }
            handlePick={[Function]}
          />,
          <button
            className="resetButton"
            onClick={[Function]}
          >
            Reset
          </button>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "",
            "className": "winnerText",
          },
          "ref": null,
          "rendered": "",
          "type": "h2",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "cards": Array [
              null,
              null,
              null,
              null,
              null,
              null,
              null,
              null,
              null,
            ],
            "handlePick": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Reset",
            "className": "resetButton",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Reset",
          "type": "button",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
